src/union_find_struct.jl

```diff
@@ -0,0 +1,22 @@
+struct UnionFind
+    rep::Dict{Int, Int} # maps each element to its representative
+end
+
+UnionFind(elems::Vector{Int}) = UnionFind(Dict{Int, Int}(x => x for x in elems))
+
+function get_rep(uf::UnionFind, x::Int)
+    while uf.rep[x] != x
+        uf.rep[x] = uf.rep[uf.rep[x]] # path compression
+        x = uf.rep[x]
+    end
+    return x
+end
+
+function union!(uf::UnionFind, x::Int, y::Int)
+    rootX = get_rep(uf, x)
+    rootY = get_rep(uf, y)
+    if rootX != rootY
```

Comment:
so here you are re-using the bus number to also represent a rank? what do you think about assigning a new_root first as min(...) and then using it to avoid calling min twice?